### PR TITLE
Switch to pybind11 style registration function API.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36258 Switch to pybind11 style registration function API.**

Previous we had a && chaining style API.  There are some downsides to
this API:

- It's easy to forget the 'static' qualifier in front, leading to
  subtle ODR bugs.
- It is not compatible with torchbind class_ definitions, as these
  need multiple levels of chaining.  So in practice people end
  up having to define multiple static initializers, one per class.
- It's not like pybind11.
- There's no way to conveniently get the file and line number of
  the registration, as there is no macro point in the API.
- The old API doesn't really encourage people to put all of their
  definitions for a library in one place, and to give a custom
  namespace for it.  Similarly, the old API wasn't very DRY, because
  you had to keep repeating the namespace/dispatch key you
  were writing implementations for.

The new API is modeled exactly off of the PYBIND11_MODULE macro:
you write:

```
TORCH_LIBRARY(aten, m) {
  m.def("aten::add(Tensor self, Tensor other) -> Tensor");
  ...
}
```

in a non-chaining fashion, and under the hood the macro expands to
define a function, and define a static initializer that allocates
c10::Library (previously called c10::Module, but we renamed it
to avoid confusion with the existing NN module concept), passes
it to your function, and then retains it for the rest of the lifetime
of the program.  Specification of the namespace is mandatory,
and in later commit I plan to make it a hard error to TORCH_LIBRARY
the same library name twice.

If you are specifying an implementation for an existing operator
(e.g., you're the XLA backend, or even if you're just putting
registrations for implementations at the implementation site),
you should use TORCH_LIBRARY_IMPL, which instead takes a backend
argument (instead of namespace) and can be used to specify an
implementation for a backend.  Unlike TORCH_LIBRARY, you can do
as many of these as you want for a backend.

This needs updates to the mobile code analyzer.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20929257](https://our.internmc.facebook.com/intern/diff/D20929257)

### CODE REVIEW

File: aten/src/ATen/autocast_mode.cpp [hunk 1]
```diff
@@ -360,23 +357,24 @@ Therefore, for the moment, this is all copy pasted in from VariableTypeEverythin
 // Common cases where registration signature matches redispatch signature
 // (that's why SIGNATURE is repeated in the WrapFunction instantiation)
 #define KERNEL(FUNC, REGISTER_NAME, SIGNATURE, POLICY) \
-  .impl(REGISTER_NAME, DispatchKey::Autocast, \
-    &WrapFunction<CastPolicy::POLICY, SIGNATURE, SIGNATURE, &FUNC>::type::call)
+  m.impl(REGISTER_NAME, \
```
Comment (bhosmer) [inline aten/src/ATen/autocast_mode.cpp:360]: Another upside to moving away from chaining: #defines are less of a bank shot.

File: aten/src/ATen/common_with_cwrap.py [hunk 2]
```diff
@@ -21,6 +21,8 @@ def parse_arguments(args):
 
 def set_declaration_defaults(declaration):
     if 'schema_string' not in declaration:
+        # This happens for legacy TH bindings like
+        # _thnn_conv_depthwise2d_backward
```
Comment (bhosmer) [inline aten/src/ATen/common_with_cwrap.py:25]: ❤️

File: aten/src/ATen/core/op_registration/op_registration.cpp [hunk 3]
```diff
@@ -171,10 +174,11 @@ Module& Module::_impl(const char* name_str, CppFunction&& f) & {
   return *this;
 }
 
-Module& Module::_fallback(CppFunction&& f) & {
-  TORCH_CHECK(!ns_, "Cannot define fallbacks from namespaces, use c10::import().fallback() instead");
-  TORCH_CHECK(f.dispatch_key_, "Fallback for catch all function not supported");
-  registrars_.emplace_back(Dispatcher::singleton().registerFallback(*f.dispatch_key_, std::move(f.func_)));
+Library& Library::_fallback(CppFunction&& f) & {
+  TORCH_CHECK(!ns_, "Cannot define a fallback in TORCH_LIBRARY (it would affect operators outside of your library); did you mean to use TORCH_LIBRARY_IMPL instead");
```
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/op_registration.cpp:178]: I found the "did you mean" part kind of confusing - it sounded like maybe there was a scoped version of fallback you could somehow specify using TORCH_LIBRARY_IMPL. 

I'm not sure how to rephrase it to more clearly connote "maybe you don't reallllly want a fallback", but absent a rephrasing maybe it's best to leave it off entirely?

File: aten/src/ATen/core/op_registration/op_registration.h [hunk 4]
```diff
@@ -605,23 +605,22 @@ class CAFFE2_API RegisterOperators final {
 //
 // A quick tour of a few usage examples:
 //
-//  auto register = torch::import()
-//
+//  // Specify a namespace.  This namespace must not be used by any
```
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/op_registration.h:608]: bikeshed: "...and populate it with all its operator definitions. (Implementations can be added here or separately.)"

File: aten/src/ATen/core/op_registration/op_registration.h [hunk 5]
```diff
@@ -774,45 +777,50 @@ namespace detail {
...
-// constructed using "torch::import".
+namespace detail {
+  class TorchLibraryInit;
+}
+
+// This is the "handle" by which functions defined in TORCH_LIBRARY
+// and TORCH_LIBRARY_IMPL can define operators and override implementations
+// at certain backends.
+//
+// Conventionally, you get access to it using those two macros:
```
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/op_registration.h:788]: Super nit: I was looking for an example of the second macro because of the ":"

File: aten/src/ATen/core/op_registration/op_registration.h [hunk 6]
```diff
@@ -832,143 +840,110 @@ class CAFFE2_API Module final {
...
   // goes way down
-  template <typename Dispatch, typename Func>
-  Module& impl_UNBOXED(const char* name, Dispatch&& key, Func* raw_f) & {
-    return impl(name, dispatch(std::forward<Dispatch>(key), CppFunction::makeUnboxedOnly(raw_f)));
-  }
-  template <typename Dispatch, typename Func>
-  Module&& impl_UNBOXED(const char* name, Dispatch&& key, Func* raw_f) && {
-    impl_UNBOXED(name, std::forward<Dispatch>(key), raw_f);
-    return std::move(*this);
+  template <typename Func>
```
Comment (bhosmer) [inline aten/src/ATen/core/op_registration/op_registration.h:888]: Super duper nit: s/these overloads/this overload/ in the comment above

File: aten/src/ATen/templates/TypeDefault.cpp [hunk 7]
```diff
@@ -29,8 +29,7 @@ namespace TypeDefault {
 
 }  // namespace TypeDefault
 
-namespace {
-auto registerer = torch::import()
+TORCH_LIBRARY(aten, m) {
```
Comment (bhosmer) [inline aten/src/ATen/templates/TypeDefault.cpp:32]: I think the invariant that this is a complete declaration of the library is a *big* improvement in quickly grokking what you're looking at.

**Review by bhosmer — APPROVED** [review]
Snazzy!

File: aten/src/ATen/templates/TypeDefault.cpp (hunk 7 above)
Comment (ezyang) [reply aten/src/ATen/templates/TypeDefault.cpp:32]: Note that the invariant is not enforced yet

File: aten/src/ATen/autocast_mode.cpp [hunk 8]
```diff
@@ -360,142 +357,144 @@ Therefore, for the moment, this is all copy pasted in from VariableTypeEverythin
...
-  KERNEL(ADD_NS(addbmm), "aten::addbmm", Tensor (const Tensor &, const Tensor &, const Tensor &, Scalar, Scalar), fp16)
-  KERNEL(ADD_NS(baddbmm), "aten::baddbmm", Tensor (const Tensor &, const Tensor &, const Tensor &, Scalar, Scalar), fp16)
-  KERNEL(ADD_NS(bmm), "aten::bmm", Tensor (const Tensor &, const Tensor &), fp16)
-  KERNEL_UNBOXED_ONLY(ADD_NS(chain_matmul), "aten::chain_matmul", Tensor (TensorList), fp16)
+TORCH_LIBRARY_IMPL(_, Autocast, m) {
+  m.fallback(c10::CppFunction::makeFallthrough());
+}
+
+TORCH_LIBRARY_IMPL(aten, Autocast, m) {
+  KERNEL_UNBOXED_ONLY(ADD_NS(_convolution), "_convolution", Tensor (const Tensor &, const Tensor &, const Tensor &, IntArrayRef, IntArrayRef, IntArrayRef, bool, IntArrayRef, int64_t, bool, bool, bool), fp16)
```
Comment (smessmer) [inline aten/src/ATen/autocast_mode.cpp:380]: do we still want these macros or can we remove it now that the API is simpler?

File: aten/src/ATen/core/op_registration/op_registration.h [hunk 9]
```diff
@@ -832,143 +863,107 @@ class CAFFE2_API Module final {
...
-//
-// So for now, we give this a scary internal name to discourage people
-// from using it
+namespace detail {
 
-// Return the namespace corresponding to the string 'ns'
-inline Module _import_DOES_NOT_WORK_WITH_MOBILE_CUSTOM_BUILD(std::string ns) {
-  return Module(std::move(ns));
-}
+class TorchLibraryInit {
```
Comment (smessmer) [inline aten/src/ATen/core/op_registration/op_registration.h:929]: nit: final

**Review by smessmer — APPROVED** [review]

File: aten/src/ATen/autocast_mode.cpp (hunk 8 above)
Comment (ezyang) [reply aten/src/ATen/autocast_mode.cpp:380]: we can remove it but I'm going to do it in another Pr

File: aten/src/ATen/core/op_registration/op_registration.h [hunk 10]
```diff
@@ -774,45 +786,55 @@ namespace detail {
...
-  Module(const Module&) = delete;
-  Module& operator=(const Module&) = delete;
+  // Which type of macro produced this Library
+  enum Kind {
+    DEF, // from TORCH_LIBRARY (no qualifier)
+    IMPL,
+    FRAGMENT,
+  };
+
+  // Use TORCH_LIBRARY/TORCH_LIBRARY_IMPL instead of these constructors directly
```
Comment (ljk53) [inline aten/src/ATen/core/op_registration/op_registration.h:828]: Note to myself: when Library constructor is used directly the current analyzer will fail.

File: aten/src/ATen/templates/TypeDefault.cpp [hunk 11]
```diff
@@ -29,8 +29,7 @@ namespace TypeDefault {
 
 }  // namespace TypeDefault
 
-namespace {
-auto registerer = torch::import()
```
Comment (ljk53) [inline aten/src/ATen/templates/TypeDefault.cpp:33]: I think you need update aten/src/ATen/templates/PerOpRegistration.cpp as well for internal buck build.

**Review by ljk53 — APPROVED** [review]

File: aten/src/ATen/core/op_registration/op_registration.h (hunk 10 above)
Comment (ezyang) [reply aten/src/ATen/core/op_registration/op_registration.h:828]: Only done in tests right now

**Comment by ezyang** [comment]
@bhosmer the error messages are all line wrapped now
